pyomo/contrib/mindtpy/feasibility_pump.py

```diff
@@ -0,0 +1,251 @@
+                        RangeSet, Constraint, Block, sqrt, TransformationFactory, ComponentMap, value)
+from pyomo.core.base.constraint import ConstraintList
+from pyomo.opt import SolverFactory, SolutionStatus
+from pyomo.contrib.gdpopt.util import SuppressInfeasibleWarning, _DoNothing, get_main_elapsed_time, copy_var_list_values, is_feasible
+from pyomo.contrib.mindtpy.nlp_solve import (solve_subproblem,
+                                             handle_subproblem_optimal, handle_subproblem_infeasible,
+                                             handle_subproblem_other_termination)
+from pyomo.contrib.mindtpy.cut_generation import add_oa_cuts, add_nogood_cuts
+from pyomo.opt import TerminationCondition as tc
+from pyomo.contrib.mindtpy.util import generate_norm2sq_objective_function
+from pyomo.contrib.mindtpy.mip_solve import solve_master, handle_master_optimal
+from pyomo.util.infeasible import log_infeasible_constraints
+
+from pyomo.contrib.mindtpy.util import generate_norm1_norm_constraint
+
+
+def feas_pump_converged(solve_data, config, discrete_only=True):
+    """Calculates the euclidean norm between the discretes in the mip and nlp models"""
+    distance = (max((nlp_var.value - milp_var.value)**2
+                    for (nlp_var, milp_var) in
+                    zip(solve_data.working_model.MindtPy_utils.variable_list,
+                        solve_data.mip.MindtPy_utils.variable_list)
+                    if (not discrete_only) or milp_var.is_integer()))
+    #
+    return distance <= config.fp_projzerotol
+
+
+def solve_feas_pump_subproblem(solve_data, config):
+    """
+    Solves the feasibility pump NLP
+
+    This function sets up the 'fp_nlp' by relax integer varibales.
+    precomputes dual values, deactivates trivial constraints, and then solves NLP model.
+
+    Parameters
+    ----------
+    solve_data: MindtPy Data Container
+        data container that holds solve-instance data
+    config: ConfigBlock
+        contains the specific configurations for the algorithm
+
+    Returns
+    -------
+    fp_nlp: Pyomo model
+        fixed NLP from the model
+    results: Pyomo results object
+        result from solving the fixed NLP
+    """
+
+    fp_nlp = solve_data.working_model.clone()
+    MindtPy = fp_nlp.MindtPy_utils
+    config.logger.info('FP-NLP %s: Solve feasibility pump NLP subproblem.'
+                       % (solve_data.fp_iter,))
+
+    # Set up NLP
+    main_objective = next(
+        fp_nlp.component_data_objects(Objective, active=True))
+    main_objective.deactivate()
+    if main_objective.sense == 'minimize':
+        fp_nlp.improving_objective_cut = Constraint(
+            expr=fp_nlp.MindtPy_utils.objective_value <= solve_data.UB)
+    else:
+        fp_nlp.improving_objective_cut = Constraint(
+            expr=fp_nlp.MindtPy_utils.objective_value >= solve_data.LB)
+
+    # Add norm_constraint, TODO: rename norm_constraint to like improving_distance_cut
```

Comment:
Is this TODO still valid?

Reply:
Yep. Do you think norm_constraint a good name? Or change it to improving_distance_cut?

Reply:
norm_constraint with documentation is good